minorminer/utils/feasibility.py

```diff
@@ -0,0 +1,222 @@
+    and this filter is designed to be efficient.
+
+    The degree distribution test is a heuristic; stronger filters are possible
+    by exploiting additional graph structure.
+
+    Args:
+        S: The source graph to embed.
+        T: The target graph in which to embed.
+        one_to_one: If True, only 1-to-1 (subgraph) embeddings are allowed.
+            Defaults to False, permitting minor embeddings.
+
+    Returns:
+        bool: `False` if subgraph embedding is definitely infeasible, `True`
+            otherwise.
+    """
+    if T == S:
+        return True
+    if (
+        T.number_of_nodes() < S.number_of_nodes()
+        or T.number_of_edges() < S.number_of_edges()
+    ):
+        return False
+    else:
+        S_degree = np.sort([S.degree[n] for n in S.nodes()])
+        T_degree = np.sort([T.degree[n] for n in T.nodes()])
+
+        if np.any(T_degree[-len(S_degree) :] < S_degree):
+            if one_to_one or T_degree[-1] <= 2:
+                # Too many high degree nodes in S
+                return False
+            else:  # Attempt minor embed (enhance T degrees)
+                # Minor embedding feasibility reduces to bin packing when
+                # considering a best target case knowing only the degree
+                # distribution. In general feasibility is NP-complete, a cheap
+                # marginal degree distribution filter is used.
+
+                # We can eliminate nodes of equal degree assuming best case:
+                ResidualCounts = Counter(T_degree)
+                ResidualCounts.subtract(Counter(S_degree))
+
+                # Target nodes of degree x <=2 are only of use for minor
+                # embedding source nodes of degree <=x:
+                for kS in range(3):
+                    if ResidualCounts[kS] < 0:
+                        ResidualCounts[kS + 1] += ResidualCounts[kS]
+                    ResidualCounts[kS] = 0
+
+                if all(v > 0 for v in ResidualCounts.values()):
+                    return True
+                nT_auxiliary = sum(ResidualCounts.values())
+                if nT_auxiliary < 0:  # extra available to form chains
+                    return False
+
+                # In best case all target nodes have degree kTmax, and chains
+                # are trees. To cover degree k in S requires n auxiliary target
+                # nodes such that kTmax + n(kTmax-2) >= k
+                kTmax = np.max([k for k, v in ResidualCounts.items() if v > 0])
+                min_auxiliary_necessary = sum(
+                    [
+                        -v * np.ceil((k - kTmax) / (kTmax - 2))
+                        for k, v in ResidualCounts.items()
+                        if v < 0
+                    ]
+                )
+                return min_auxiliary_necessary <= nT_auxiliary
+        else:
+            return True
+
+
+def lattice_size_upper_bound(T: nx.Graph = None) -> int:
```

Comment:
If T can actually be `None` it should both be typed as `Optional[nx.Graph]` and fixed so that the code doesn't fail. Alt., I suspect that it's better just to remove the default.
```suggestion
def lattice_size_upper_bound(T: nx.Graph) -> int:
```

Reply:
I'm deeply uncomfortable with the name of this function.  I expected it to have the same signature of `lattice_size_lower_bound` and perform similar analysis.  But as I write that out, I'm not sure how I would implement such a function in the `one_to_one` case.  It seems Hard.

Reply:
I guess, not that hard -- `lattice_size_upper_bound` could just have formulae for clique numbers of perfectly yielded lattices:
Chimera: M*T
Pegasus: 12*M-12
Zephyr: 4*T*M - 2*T

Reply:
Well, that's the embedding case -- for `one_to_one`, I suspect that the diameter of the graph should be an upper bound, but that only be asymptotically true.

Reply:
Maybe best to rename this slightly (to diff it from `lattice_size_lower_bound`). We can work on adding a tighter upper bound calculation later.